pkg/controller/jobs/leaderworkerset/leaderworkerset_reconciler.go

```diff
@@ -0,0 +1,227 @@
+	ctrl.Log.V(3).Info("Setting up LeaderWorkerSet reconciler")
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&leaderworkersetv1.LeaderWorkerSet{}).
+		Named("leaderworkerset").
+		WithEventFilter(r).
+		Complete(r)
+}
+
+// +kubebuilder:rbac:groups=leaderworkerset.x-k8s.io,resources=leaderworkersets,verbs=get;list;watch
+
+func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	lws := &leaderworkersetv1.LeaderWorkerSet{}
+	err := r.client.Get(ctx, req.NamespacedName, lws)
+	if err != nil {
+		// we'll ignore not-found errors, since there is nothing to do.
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	log := ctrl.LoggerFrom(ctx).WithValues("leaderworkerset", klog.KObj(lws))
+	ctx = ctrl.LoggerInto(ctx, log)
+	log.V(2).Info("Reconciling LeaderWorkerSet")
+
+	err = r.createPrebuildWorkloadsIfNotExist(ctx, lws)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *Reconciler) createPrebuildWorkloadsIfNotExist(ctx context.Context, lws *leaderworkersetv1.LeaderWorkerSet) error {
+	replicas := ptr.Deref(lws.Spec.Replicas, 1)
+	for i := int32(0); i < replicas; i++ {
+		if err := r.createPrebuildWorkloadIfNotExist(ctx, lws, i); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *Reconciler) createPrebuildWorkloadIfNotExist(ctx context.Context, lws *leaderworkersetv1.LeaderWorkerSet, index int32) error {
+	wl := &kueue.Workload{}
+	err := r.client.Get(ctx, client.ObjectKey{Name: GetWorkloadName(lws.UID, lws.Name, fmt.Sprint(index)), Namespace: lws.Namespace}, wl)
+	// Ignore if the Workload already exists or an error occurs.
+	if err == nil || client.IgnoreNotFound(err) != nil {
+		return err
+	}
+	if err := r.createPrebuildWorkload(ctx, lws, index); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *Reconciler) createPrebuildWorkload(ctx context.Context, lws *leaderworkersetv1.LeaderWorkerSet, index int32) error {
+	createdWorkload := r.constructWorkload(lws, index)
+	err := r.client.Create(ctx, createdWorkload)
+	if err != nil {
+		return err
+	}
+	r.record.Eventf(
+		lws, corev1.EventTypeNormal, jobframework.ReasonCreatedWorkload,
+		"Created Workload: %v", workload.Key(createdWorkload),
+	)
+	return nil
+}
+
+func (r *Reconciler) constructWorkload(lws *leaderworkersetv1.LeaderWorkerSet, index int32) *kueue.Workload {
+	wl := &kueue.Workload{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        GetWorkloadName(lws.UID, lws.Name, fmt.Sprint(index)),
+			Namespace:   lws.Namespace,
+			Finalizers:  []string{kueue.ResourceInUseFinalizerName},
+			Annotations: admissioncheck.FilterProvReqAnnotations(lws.GetAnnotations()),
+		},
+		Spec: kueue.WorkloadSpec{
+			QueueName:                   jobframework.QueueNameForObject(lws),
+			PodSets:                     r.podSets(lws),
+			MaximumExecutionTimeSeconds: jobframework.MaximumExecutionTimeSecondsForObject(lws),
+		},
+	}
```

Comment:
Can we commonize this more? Note that this requires duplication for `MaximumExecutionTimeSeconds`, but it might also require duplication for other fields in the future, and we are likely to forget about this function here.

Reply:
Done on a351825.